### PR TITLE
[flang] Add struct passing target rewrite hooks and partial X86-64 impl

### PR DESCRIPTION
**Pull request by jeanPerier**
In the context of C/Fortran interoperability (BIND(C)), it is possible to give the VALUE attribute to a BIND(C) derived type dummy, which according to Fortran 2018 18.3.6 - 2. (4) implies that it must be passed like the equivalent C structure value. The way C structure value are passed is ABI dependent.

LLVM does not implement the C struct ABI passing for LLVM aggregate type arguments. It is up to the front-end, like clang is doing, to split the struct into registers or pass the struct on the stack (llvm "byval") as required by the target ABI.
So the logic for C struct passing sits in clang. Using it from flang requires setting up a lot of clang context and to bridge FIR/MLIR representation to clang AST representation for function signatures (in both directions). It is a non trivial task.
See https://stackoverflow.com/questions/39438033/passing-structs-by-value-in-llvm-ir/75002581#75002581.

Since BIND(C) struct are rather limited as opposed to generic C struct (e.g. no bit fields). It is easier to provide a limited implementation of it for the case that matter to Fortran.

This patch:
- Updates the generic target rewrite pass to keep track of both the new argument type and attributes. The motivation for this is to be able to tell if a previously marshalled argument is passed in memory (it is a C pointer), or if it is being passed on the stack (has the byval llvm attributes).
- Adds an entry point in the target specific codegen to marshal struct arguments, and use it in the generic target rewrite pass.
- Implements limited support for the X86-64 case. So far, the support allows telling if a struct must be passed in register or on the stack, and to deal with the stack case. The register case is left TODO in this patch.

The X86-64 ABI implemented is the System V ABI for AMD64 version 1.0

### CODE REVIEW

File: flang/lib/Optimizer/CodeGen/Target.cpp [hunk 1]
```diff
@@ -58,6 +59,62 @@ static void typeTodo(const llvm::fltSemantics *sem, mlir::Location loc,
...
+    return {size, align};
+  }
+  if (auto logical = mlir::dyn_cast<fir::LogicalType>(ty)) {
+    mlir::Type intTy = mlir::IntegerType::get(
+        logical.getContext(), kindMap.getLogicalBitsize(logical.getFKind()));
+    return getSizeAndAlignment(loc, intTy, dl, kindMap);
+  }
+  if (auto logical = mlir::dyn_cast<fir::CharacterType>(ty)) {
+    mlir::Type intTy = mlir::IntegerType::get(
+        logical.getContext(), kindMap.getLogicalBitsize(logical.getFKind()));
```
Comment (tblah) [inline flang/lib/Optimizer/CodeGen/Target.cpp:112]:
```suggestion
  if (auto character = mlir::dyn_cast<fir::CharacterType>(ty)) {
    mlir::Type intTy = mlir::IntegerType::get(
        character.getContext(), kindMap.getLogicalBitsize(character.getFKind()));
```

**Review by tblah — COMMENTED** [review]
The approach looks good. I don't know enough about the X86 ABI to comment on any specifics.

File: flang/lib/Optimizer/CodeGen/Target.cpp (hunk 1 above)
Comment (jeanPerier) [reply flang/lib/Optimizer/CodeGen/Target.cpp]: Thanks! Updated this to use getCharacterBitsize.

**Review by psteinfeld — CHANGES_REQUESTED** [review]
I didn't understand most of it, but what I understood looks good.

I found one nit and one compilation error that needs to be fixed.

**Comment by kiranchandramohan** [comment]
Thanks @jeanPerier for the patch. I have not had a chance to look at the patch in detail. I have two general questions,
-> Are you looking for help implementing the target specific parts for other targets?
-> The summary talks specifically about BIND(C) here, but doesn't the rules apply to derived types passed by value in general. Say a derived type that looks like a complex type internally, wouldn't that need some target specific handling like Complex?

**Comment by psteinfeld** [comment]
> > I found one nit and one compilation error that needs to be fixed.
> 
> Thanks @psteinfeld. I fixed the assert issue. What is the nit you mentionned?
In flang/lib/Optimizer/CodeGen/Target.cpp, line 399, "ystem" should read "system".

All builds and tests successfully now for me.

Thanks for the quick fix!

**Review by psteinfeld — APPROVED** [review]
Looks great!

**Review by vzakhari — COMMENTED** [review]
Thank you for working on this, Jean!

File: flang/lib/Optimizer/CodeGen/Target.cpp [hunk 2]
```diff
@@ -318,6 +382,251 @@ struct TargetX86_64 : public GenericTarget<TargetX86_64> {
...
+  // Goes through the previously marshalled arguments and count the
+  // register occupancy to check if there are enough registers left.
+  bool hasEnoughRegisters(mlir::Location loc, int neededIntRegisters,
+                          int neededSSERegisters,
+                          const Marshalling &previousArguments) const {
+    int availIntRegisters = 6;
+    int availSSERegisters = 8;
+    for (auto typeAndAttr : previousArguments) {
+      const auto &attr = std::get<Attributes>(typeAndAttr);
+      if (attr.isByVal() || attr.isSRet())
```
[the remark this answers is not in the document]
Comment (jeanPerier) [reply flang/lib/Optimizer/CodeGen/Target.cpp]: You are correct, 3.2.3 "Returning of Values" point 2 this is using %rdi. Updated.

**Review by jeanPerier — COMMENTED** [review]
Thanks for the reviews

**Review by vzakhari — APPROVED** [review]
Thank you for the update, Jean!  I have only one minor comment about `0` alignment.

File: flang/lib/Optimizer/CodeGen/Target.cpp [hunk 3]
```diff
@@ -58,6 +59,63 @@ static void typeTodo(const llvm::fltSemantics *sem, mlir::Location loc,
...
+    auto [eleSize, eleAlign] =
+        getSizeAndAlignment(loc, seqTy.getEleTy(), dl, kindMap);
+
+    std::uint64_t size =
+        llvm::alignTo(eleSize, eleAlign) * seqTy.getConstantArraySize();
+    return {size, eleAlign};
+  }
+  if (auto recTy = mlir::dyn_cast<fir::RecordType>(ty)) {
+    std::uint64_t size = 0;
+    unsigned short align = 0;
```
[the remark this answers is not in the document]
Comment (jeanPerier) [reply flang/lib/Optimizer/CodeGen/Target.cpp]: Yes, that is safer, thanks.

**Pull request merged**